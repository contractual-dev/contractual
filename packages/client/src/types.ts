import type { ApiOperations } from '../client/index.js';
import type { ApiClient } from './api-client';

export type ApiOperationToClientMethod<TOperation extends keyof typeof ApiOperations> =
  (typeof ApiOperations)[TOperation];

export type ApiClientInput<TMethod extends keyof ApiClient> = ApiClient[TMethod] extends (
  data: infer TInput
) => any
  ? TInput
  : never;
