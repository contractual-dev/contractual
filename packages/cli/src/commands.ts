import { Command } from 'commander';
import { generateContract, generateSpec } from './commands/generate.command.js';
import inquirer from 'inquirer';

const program = new Command();
program.name('contractual').description('A sample CLI tool');

const generateCommand = new Command('generate').description('Generate resources');

generateCommand
  .command('contract')
  .description('Generate a contract based on the provided OpenAPI file')
  .action(() => {
    return generateContract();
  });

generateCommand
  .command('spec')
  .description('Generate spec')
  .action(async () => {
    await generateSpec(inquirer);
  });

program.addCommand(generateCommand);

program.parse(process.argv);
