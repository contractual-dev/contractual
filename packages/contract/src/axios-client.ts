import type { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import axios from 'axios';
import * as crypto from 'node:crypto';

interface LoggedRequest {
  id: string;
  method: string;
  url: string;
  headers: Record<string, string>;
  body?: any;
}

interface LoggedResponse {
  id: string;
  status: number;
  data: any;
  headers: Record<string, string>;
  duration: number; // In milliseconds
}

// Use a Map to store metadata per request
const requestMetadata = new Map<string, { startTime: Date }>();

const hashUniqueRequest = (operation: AxiosRequestConfig): string => {
  const hash = crypto.createHash('sha256');

  hash.update(
    JSON.stringify({
      method: operation.method,
      url: operation.url,
      headers: operation.headers,
      body: operation.data,
      params: operation.params,
    })
  );

  return hash.digest('hex');
};

export const createAxiosClient = (): { client: AxiosInstance; logs: any[] } => {
  const logs: Array<{ request: LoggedRequest; response: LoggedResponse }> = [];
  const client = axios.create();

  // Add a request interceptor
  client.interceptors.request.use(
    (config) => {
      const requestId = hashUniqueRequest(config);

      // Store metadata for the request in the Map
      requestMetadata.set(requestId, { startTime: new Date() });

      return config;
    },
    (error) => Promise.reject(error)
  );

  client.interceptors.response.use(
    (response: AxiosResponse) => {
      const config = response.config;
      const requestId = hashUniqueRequest(config);

      const metadata = requestMetadata.get(requestId);

      if (metadata) {
        const { startTime } = metadata;
        const duration = new Date().getTime() - startTime.getTime();

        // Log request and response
        logs.push({
          request: {
            id: requestId,
            method: response.config.method?.toUpperCase() || '',
            url: response.config.url || '',
            headers: response.config.headers as Record<string, string>,
            body: response.config.data,
          },
          response: {
            id: requestId,
            status: response.status,
            data: response.data,
            headers: response.headers as Record<string, string>,
            duration,
          },
        });

        // Clean up metadata for the completed request
        requestMetadata.delete(requestId);
      }

      return response;
    },
    (error) => {
      if (error.config) {
        const config = error.config as any;
        const requestId = config.requestId; // Retrieve the requestId from config

        const metadata = requestMetadata.get(requestId);

        if (metadata) {
          const { startTime } = metadata;
          const duration = new Date().getTime() - startTime.getTime();

          logs.push({
            request: {
              id: requestId,
              method: error.config.method?.toUpperCase() || '',
              url: error.config.url || '',
              headers: error.config.headers as Record<string, string>,
              body: error.config.data,
            },
            response: {
              id: requestId,
              status: error.response?.status,
              data: error.response?.data,
              headers: error.response?.headers || {},
              duration,
            },
          });

          // Clean up metadata for the failed request
          requestMetadata.delete(requestId);
        }
      }

      return Promise.reject(error);
    }
  );

  return { client, logs };
};
