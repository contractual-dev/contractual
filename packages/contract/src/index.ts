import * as path from 'node:path';
export * from './api-client.js';
export * from './types.js';

export function getClientPath() {
  return path.resolve('..', 'client');
}
