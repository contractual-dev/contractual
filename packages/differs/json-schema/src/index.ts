import { diffSchemas } from 'json-schema-diff';

export type schema = { [key: string]: any };

export function createJSONSchemaDiffer() {
  const api = {
    diff(sourceSchema: any, destinationSchema: any) {
      console.log('hello world')
      try {
        return diffSchemas({
          sourceSchema,
          destinationSchema,
        });
      } catch (e) {
        console.error('error found!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!')
      }
    },
  };

  return api;
}
