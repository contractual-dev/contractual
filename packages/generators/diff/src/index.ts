export * from './generator.js';
export * from './diff-highlighter.js';
