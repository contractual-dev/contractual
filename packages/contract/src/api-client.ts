import type { InitClientReturn, ClientArgs } from '@ts-rest/core';
import { initClient } from '@ts-rest/core';
import type { AxiosError, AxiosResponse, Method } from 'axios';
import { isAxiosError } from 'axios';
import { ApiContract } from '../contract/index.js';
import { createAxiosClient } from './axios-client.js';

export type ApiClient = ReturnType<typeof getApiClient>;

let apiClientInstance: InitClientReturn<typeof ApiContract, ClientArgs> | null = null;

export function getApiClient(options: {
  baseUrl: string;
  baseHeaders: Record<string, string> | undefined;
}) {
  if (!apiClientInstance) {
    apiClientInstance = initClient(ApiContract, {
      baseUrl: options.baseUrl,
      baseHeaders: options.baseHeaders,
      api: async ({ path, method, headers, body }) => {
        try {
          const axiosClient = createAxiosClient().client;

          const result = await axiosClient.request({
            method: method as Method,
            url: path,
            headers,
            data: body,
          });

          return {
            status: result.status,
            body: result.data,
            headers: new Headers(Object.entries(result.headers ?? {})),
          };
        } catch (error: Error | AxiosError | any) {
          if (isAxiosError(error)) {
            const err = error as AxiosError;
            const response = err.response as AxiosResponse;

            return {
              status: response.status,
              body: response.data,
              headers: new Headers(Object.entries(response.headers ?? {})),
            };
          }

          throw error;
        }
      },
    });
  }

  return apiClientInstance;
}
