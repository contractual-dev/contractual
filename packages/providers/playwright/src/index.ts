import type { apiRouter } from '@contractual/contract/contract/index.js';
import { type ApiClient, type ApiClientInput, getApiClient } from '@contractual/contract';
import { ApiOperations } from '@contractual/contract/contract/index.js';
import { Fixtures } from '@contractual/fixtures/fixtures/index.js';
import type { TestFunctionParams } from '@contractual/types.test';
import { test as playwrightTest } from '@playwright/test';
import type { ClientArgs, InitClientReturn } from '@ts-rest/core';

const testFixture = playwrightTest.extend<{ client: ApiClient }>({
  client: async ({ baseURL, extraHTTPHeaders }, use) => {
    await use(
      getApiClient({
        baseUrl: baseURL || 'http://localhost:3000',
        baseHeaders: extraHTTPHeaders,
      })
    );
  },
});

export function test<TOperation extends keyof typeof ApiOperations>(
  operation: TOperation,
  fn: (fixtures: TestFunctionParams<TOperation>) => Promise<void>
) {
  // Create the operation function
  const operationFn = (
    clientMethod: (typeof ApiOperations)[TOperation],
    client: InitClientReturn<typeof apiRouter, ClientArgs>
  ) =>
    Object.assign(
      <TFixture extends keyof (typeof Fixtures)[TOperation]>(fixture: TFixture) => {
        const resolvedFixture = (Fixtures as Record<TOperation, never>)[operation][fixture];

        return (
          client[clientMethod] as (
            args: ApiClientInput<typeof clientMethod>
          ) => ReturnType<ApiClient[typeof clientMethod]>
        )(resolvedFixture as ApiClientInput<typeof clientMethod>);
      },
      (cb?: () => ApiClientInput<typeof clientMethod>) => {
        if (!cb) {
          throw new Error('Callback function is required for operation');
        }

        return (
          client[clientMethod] as (
            args: ApiClientInput<typeof clientMethod>
          ) => ReturnType<ApiClient[typeof clientMethod]>
        )(cb());
      }
    );

  // Extend Playwright's test runner
  return testFixture(operation, async ({ client }) => {
    // Resolve the client method
    const clientMethod = ApiOperations[operation];

    // Call the user-defined function with resolved fixtures and test info
    await fn({ operation: operationFn(clientMethod, client) });
  });
}
