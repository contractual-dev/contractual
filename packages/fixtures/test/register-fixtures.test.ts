import type { RegisterFixturesReturnType } from '../src';
import { registerFixtures } from '../src';

describe('Register Fixtures Unit Spec', () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-expect-error
  let result: RegisterFixturesReturnType<'create version'>;

  beforeAll(() => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    result = registerFixtures('create version', {
      'basic version': () => ({
        body: {
          version: 'string',
          codename: 'string',
          from: 'string',
          is_stable: true,
          is_beta: true,
          is_hidden: true,
          is_deprecated: true,
        },
      }),
      'rc version': ({ extend }) => {
        return extend({
          'with something else': () => ({
            body: {
              version: 'string',
              from: 'string',
            },
          }),
        });
      },
    });
  });

  it('should return the correct snapshot', () => {
    expect(result).toMatchSnapshot();
  });
});
