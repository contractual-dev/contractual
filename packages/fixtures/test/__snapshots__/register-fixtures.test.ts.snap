// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Fixtures Unit Spec should return the correct snapshot 1`] = `
{
  "fixtures": {
    "basic version": {
      "body": {
        "codename": "string",
        "from": "string",
        "is_beta": true,
        "is_deprecated": true,
        "is_hidden": true,
        "is_stable": true,
        "version": "string",
      },
    },
    "rc version with something else": {
      "body": {
        "from": "string",
        "version": "string",
      },
    },
  },
  "operation": "create version",
}
`;
